include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -1,44 +1,140 @@
+#endif
+}
 
 namespace ray_dir_info
 {
 
-// no ray-differentials, nothing
-struct Basic
+#define NBL_CONCEPT_NAME Basic
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (rdirinfo, T)
+#define NBL_CONCEPT_PARAM_1 (N, typename T::vector3_type)
+#define NBL_CONCEPT_PARAM_2 (dirDotN, typename T::scalar_type)
+#define NBL_CONCEPT_PARAM_3 (m, typename T::matrix3x3_type)
+NBL_CONCEPT_BEGIN(4)
+#define rdirinfo NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+#define N NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_1
+#define dirDotN NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_2
+#define m NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_3
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_TYPE)(T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::vector3_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::matrix3x3_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.direction), ::nbl::hlsl::is_same_v, typename T::vector3_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.getDirection()), ::nbl::hlsl::is_same_v, typename T::vector3_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.transmit()), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.reflect(N, dirDotN)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.refract(N, dirDotN)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::transform(m, rdirinfo)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(is_scalar_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(is_vector_v, typename T::vector3_type))
+);
+#undef m
+#undef dirDotN
+#undef N
+#undef rdirinfo
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
+
+template <typename T>
+struct SBasic
 {
-  float3 getDirection() {return direction;}
-
-  Basic transmit()
-  {
-    Basic retval;
-    retval.direction = -direction;
-    return retval;
-  }
-  
-  Basic reflect(const float3 N, const float directionDotN)
-  {
-    Basic retval;
-    retval.direction = nbl::hlsl::reflect(direction,N,directionDotN);
-    return retval;
-  }
-
-  float3 direction;
+    using scalar_type = T;
+    using vector3_type = vector<T, 3>;
+    using matrix3x3_type = matrix<T, 3, 3>;
+
+    vector3_type getDirection() NBL_CONST_MEMBER_FUNC { return direction; }
+
+    SBasic<T> transmit()
+    {
+        SBasic<T> retval;
+        retval.direction = -direction;
+        return retval;
+    }
+
+    SBasic<T> reflect(NBL_CONST_REF_ARG(vector3_type) N, scalar_type directionDotN)
+    {
+        SBasic<T> retval;
+        retval.direction = bxdf::reflect<vector3_type>(direction,N,directionDotN);
+        return retval;
+    }
+
+    SBasic<T> refract(NBL_CONST_REF_ARG(vector3_type) N, scalar_type eta)
+    {
+        SBasic<T> retval;
+        bxdf::refract<vector3_type> r = bxdf::refract<vector3_type>::create(direction,N,eta);
+        retval.direction = r.doRefract();
```

Comment:
take the refract struct from outside (it should have all the `eta` and `N` as members 

because right now you're taking `N` and `eta` separately, which also means we have no opportunity to provide precomputed dot products of N with direction, etc.